File: problems/3219_minimum_cost_for_cutting_cake_ii/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    long long minimumCost(int m, int n, vector<int>& horizontalCut, vector<int>& verticalCut) {
        sort(horizontalCut.rbegin(), horizontalCut.rend());
        sort(verticalCut.rbegin(), verticalCut.rend());
        int h=0,v=0;
        long long hp=1,vp=1,res=0;
        while (h<(int)horizontalCut.size()&&v<(int)verticalCut.size()) {
            if (horizontalCut[h]>=verticalCut[v]) { res+=(long long)horizontalCut[h++]*vp; hp++; }
            else { res+=(long long)verticalCut[v++]*hp; vp++; }
        }
        while (h<(int)horizontalCut.size()) res+=(long long)horizontalCut[h++]*vp;
        while (v<(int)verticalCut.size()) res+=(long long)verticalCut[v++]*hp;
        return res;
    }
};

File: problems/3219_minimum_cost_for_cutting_cake_ii/Solution.java
import java.util.*;

class Solution {
    public long minimumCost(int m, int n, int[] horizontalCut, int[] verticalCut) {
        Arrays.sort(horizontalCut); Arrays.sort(verticalCut);
        int h = horizontalCut.length-1, v = verticalCut.length-1;
        long hp = 1, vp = 1, res = 0;
        while (h >= 0 && v >= 0) {
            if (horizontalCut[h] >= verticalCut[v]) { res += horizontalCut[h--] * vp; hp++; }
            else { res += verticalCut[v--] * hp; vp++; }
        }
        while (h >= 0) res += horizontalCut[h--] * vp;
        while (v >= 0) res += verticalCut[v--] * hp;
        return res;
    }
}
